src/rust/engine/options/src/config.rs

```diff
@@ -217,21 +217,132 @@ fn toml_table_to_dict(table: &Value) -> HashMap<String, Val> {
 }
 
 #[derive(Clone)]
-pub(crate) struct Config {
+struct ConfigSource {
+    #[allow(dead_code)]
+    path: Option<OsString>,
     config: Value,
 }
 
+impl ConfigSource {
+    fn option_name(id: &OptionId) -> String {
+        id.name("_", NameTransform::None)
+    }
+
+    fn get_value(&self, id: &OptionId) -> Option<&Value> {
+        self.config
+            .get(id.scope())
+            .and_then(|table| table.get(Self::option_name(id)))
+    }
+
+    fn get_list<T: FromValue>(
+        &self,
+        id: &OptionId,
+        parse_list: fn(&str) -> Result<Vec<ListEdit<T>>, ParseError>,
+    ) -> Result<Vec<ListEdit<T>>, String> {
+        let mut list_edits = vec![];
+        if let Some(table) = self.config.get(id.scope()) {
+            let option_name = Self::option_name(id);
+            if let Some(value) = table.get(&option_name) {
+                match value {
+                    Value::Table(sub_table) => {
+                        if sub_table.is_empty()
+                            || !sub_table.keys().collect::<HashSet<_>>().is_subset(
+                                &["add".to_owned(), "remove".to_owned()]
```

Comment:
Nit: this code is heading off far to the right, and the "happy branch" contains the rest of the function. So I'd really switch these around to `let Some(...) = ... else { return/continue };`, which would dedent at least twice and hopefully make the happy path clearer. :)

Reply:
I took a stab at changing this, but looking at it, as a matter of personal preference I think I find this more readable as-is.

Reply:
Fair enough! Thanks for looking.